### ISSUE TITLE
BUG: invalid implementation of NTOWFv2 hashing

### ISSUE THREAD

**Issue by arminfelder** [versat/cntlm#83]
according to https://winprotocoldoc.blob.core.windows.net/productionwindowsarchives/MS-NLMP/%5bMS-NLMP%5d.pdf page 59
```Python
Define NTOWFv2(Passwd, User, UserDom) as HMAC_MD5(
MD4(UNICODE(Passwd)), UNICODE(ConcatenationOf( Uppercase(User),
UserDom ) ) )
EndDefine
Define LMOWFv2(Passwd, User, UserDom) as NTOWFv2(Passwd, User,
UserDom)
EndDefine
...
```
the **userDom** has to be used for hashing the password, but cntlm instead uses the **target**, which should have the value of **userDom**, because cntlm is requesting the server to add the target information(NTLMSSP_REQUEST_TARGET flag), but there are cases e.g. McAffee Webgateway, where the request for **target**, is ignored (although this is a violation of the protocol ).

```C
	if (creds->hashntlm2 && !tblen) {
		return 0;
	}

	if (creds->hashntlm2) {
		ntlm2_calc_resp(&nthash, &ntlen, &lmhash, &lmlen, creds->passntlm2, challenge, tbofs, tblen);
	}
```

by changing cntlm to use userDom instead of target, we would be compliant with the protocol spec, and also to able handle edge cases as mentioned

something like:
```C
	if (creds->hashntlm2 && !userDomofs) {
		return 0;
	}

	if (creds->hashntlm2) {
		ntlm2_calc_resp(&nthash, &ntlen, &lmhash, &lmlen, creds->passntlm2, challenge, userDomofs, userDomlen);
	}
```

**Comment by jschwartzenberg** [comment]
It seems that PR had two issues:
1. it was quite big, trying to tackle multiple problems
2. it broke authentication in some environments

My proposal would be to have a PR with just this fix and to introduce the structs refactoring in a separate PR. Then both can be tested independently for compatibility with different implementations. As they might touch the same code, you may have to provide one after the other. Smaller PRs with independent fixes have a much higher chance than getting merged than large PRs.

**Comment by fralken** [comment]
Thanks @arminfelder. Can you provide a PR with just the fix (how to retrieve the userDom and use it instead of target), and after another PR with the code refactoring?

**Comment by fralken** [comment]
Indeed the equivalent of NTOWFv2 in cntlm is correct, because it is `passntlm2` that is computed by this function:
https://github.com/versat/cntlm/blob/83f7c54ac63f4e3bc51aedcbdc6dbe06bf3e2b20/ntlm.c#L200-L224

**Comment by fralken** [comment]
It turns out that the actual issue was that the authentication message is not complete when the target info is missing from the challenge. This is fixed with commit 533f1379af36a5a49cd6bdc074a686937ce9242e.
